File: crates/algokit_utils/src/network_client.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub enum TokenHeader {
    String(String),
    Headers(HashMap<String, String>),
}

#[derive(Debug, Clone)]
pub struct AlgoClientConfig {
    pub server: String,
    pub port: Option<u16>,
    pub token: Option<TokenHeader>,
}

impl AlgoClientConfig {
    pub fn new(server: String) -> Self {
        Self {
            server,
            port: None,
            token: None,
        }
    }

    pub fn with_port(mut self, port: u16) -> Self {
        self.port = Some(port);
        self
    }

    pub fn with_token(mut self, token: TokenHeader) -> Self {
        self.token = Some(token);
        self
    }
}

#[derive(Debug, Clone)]
pub struct AlgoConfig {
    pub algod_config: AlgoClientConfig,
}

impl AlgoConfig {
    pub fn new(algod_config: AlgoClientConfig) -> Self {
        Self { algod_config }
    }
}

#[derive(Debug, Clone)]
pub struct NetworkDetails {
    pub is_testnet: bool,
    pub is_mainnet: bool,
    pub is_localnet: bool,
    pub genesis_id: String,
    pub genesis_hash: String,
}

impl NetworkDetails {
    pub fn new(genesis_id: String, genesis_hash: String) -> Self {
        let is_localnet = genesis_id_is_localnet(&genesis_id);
        let is_testnet = genesis_id == "testnet-v1.0";
        let is_mainnet = genesis_id == "mainnet-v1.0";

        Self {
            is_testnet,
            is_mainnet,
            is_localnet,
            genesis_id,
            genesis_hash,
        }
    }
}

pub fn genesis_id_is_localnet(genesis_id: &str) -> bool {
    genesis_id == "devnet-v1" || genesis_id == "sandnet-v1" || genesis_id == "dockernet-v1"
}

pub trait AlgoClients {
    type AlgodClient;
    fn algod(&self) -> &Self::AlgodClient;
}

File: packages/typescript/algokit_utils/src/clients/network-client.ts
export type TokenHeader = string | Record<string, string>

/** Represents the different Algorand networks */
export enum AlgorandNetwork {
  /** Local development network */
  LocalNet = 'localnet',
  /** Algorand TestNet */
  TestNet = 'testnet',
  /** Algorand MainNet */
  MainNet = 'mainnet',
}

/** Represents the different Algorand services */
export enum AlgorandService {
  /** Algorand daemon (algod) - provides access to the blockchain */
  Algod = 'algod',
  /** Algorand indexer - provides historical blockchain data */
  Indexer = 'indexer',
  /** Key Management Daemon (kmd) - provides key management functionality */
  Kmd = 'kmd',
}

/** Config for an Algorand SDK client. */
export interface AlgoClientConfig {
  /** Base URL of the server e.g. http://localhost, https://testnet-api.algonode.cloud/, etc. */
  server: string
  /** Optional port to use e.g. 4001, 443, etc. */
  port?: number | string
  /** Optional token or headers to use for API authentication */
  token?: TokenHeader
}

/** Configuration for algod, indexer and kmd clients. */
export interface AlgoConfig {
  /** Algod client configuration */
  algodConfig: AlgoClientConfig
  /** Indexer client configuration */
  indexerConfig?: AlgoClientConfig
  /** Kmd client configuration */
  kmdConfig?: AlgoClientConfig
}

/** Details about the currently connected network. */
export interface NetworkDetails {
  /** Whether the network is TestNet */
  isTestnet: boolean
  /** Whether the network is MainNet */
  isMainnet: boolean
  /** Whether the network is a LocalNet */
  isLocalnet: boolean
  /** Genesis ID reported by the network */
  genesisId: string
  /** Genesis hash reported by the network encoded as base64 */
  genesisHash: string
}

/**
 * Returns true if the given network genesisId is associated with a LocalNet network.
 * @param genesisId The network genesis ID
 * @returns Whether the given genesis ID is associated with a LocalNet network
 */
export function genesisIdIsLocalNet(genesisId: string): boolean {
  return genesisId === 'devnet-v1' || genesisId === 'sandnet-v1' || genesisId === 'dockernet-v1'
}
